Script to add new rule file

acharles7 opened the pull request:
## Summary

Fixes #119 
* Added script to create `new_rule.py`
```bash
$ python -m fixit.cli.add_new_rule --path fixit/rules/new_rule.py
```

This will create a skeleton of the rule file at the path given in the `--path` argument

`new_rule.py`
```python
# Copyright (c) Facebook, Inc. and its affiliates.
#
# This source code is licensed under the MIT license found in the
# LICENSE file in the root directory of this source tree.


from fixit import CstLintRule, InvalidTestCase as Invalid, ValidTestCase as Valid


"""
This is a model rule file for adding a new rule to fixit module
"""


class Rule(CstLintRule):
    """
    docstring or new_rule description
    """

    MESSAGE = "Enter rule description message"

    VALID = [Valid("'example'")]

    INVALID = [Invalid("'example'")]
```

jimmylai commented on fixit/cli/add_new_rule.py:
```diff
@@ -0,0 +1,88 @@
+
+def create_rule_file(file_path: Path) -> None:
+    """Create a new rule file."""
+    context = _LICENCE + _IMPORTS + _TO_DOS + _RULE_CLASS
```
Let's use the formatter to reformat the code based on user's config.
https://github.com/Instagram/Fixit/blob/ef5edca7c6d3cc5f2f1cb5ff54b769ac8bb0d923/fixit/cli/apply_fix.py#L126-L128

acharles7 replied:
Done

jimmylai reviewed (approved):
Let's also include this tool in the documentation.

https://github.com/Instagram/Fixit/blob/46a0d0b66e494c7dad4fe5d57dea9981a96df322/docs/source/build_a_lint_rule.ipynb

You can run `jupyter notebook` in your local venv to start Jupyter notebook and edit it in browser.
It uses nbsphinx to generate doc and text cell needs to use `rst` format.
https://nbsphinx.readthedocs.io/en/0.7.1/raw-cells.html
The best place to mention this command is in the section of "Lint Rule Scaffolding".

acharles7 commented on fixit/cli/add_new_rule.py:
```diff
@@ -0,0 +1,88 @@
+
+"""
+
+_LICENCE = """# Copyright (c) Facebook, Inc. and its affiliates.
```
```
_LICENCE = """\
# Copyright (c) Facebook, Inc. and its affiliates.
```
this gives me lint error in windows python that's why I removed it.
Please suggest me other way to do this.

isidentical replied:
It is weird that only windows fail, and I have no idea. Guess you have to revert this.

acharles7 replied:
I'll revert it.

jimmylai replied:
Yeah, I saw the error on Windows before and didn't know how to fix it.

isidentical commented on fixit/cli/add_new_rule.py:
```diff
@@ -0,0 +1,93 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# This source code is licensed under the MIT license found in thefla
```
```suggestion
# This source code is licensed under the MIT license found in the
```

acharles7 replied:
Done

isidentical commented on fixit/cli/add_new_rule.py:
```diff
@@ -0,0 +1,93 @@
+
+"""
+
+_IMPORTS = """
```
Just my opinion, but wouldn't it be cool to have `libcst as cst` and `libcst.matchers as m` imports here as well?

acharles7 replied:
Will add those imports.

acharles7 replied:
Done

isidentical commented:
As far as I could estimate from writing 2 rules is that, saving the rule to `new_file.py` by default isn't that useful. Maybe it can take a positional argument named rule name and use it on both places (the filename and the class name, so that people wouldn't need to replace the class name, because they might just remove the `Rule` at all which is non-compliant with the style standards of this repo) (just my opinion)

acharles7 commented:
> As far as I could estimate from writing 2 rules is that, saving the rule to `new_file.py` by default isn't that useful. Maybe it can take a positional argument named rule name and use it on both places (the filename and the class name, so that people wouldn't need to replace the class name, because they might just remove the `Rule` at all which is non-compliant with the style standards of this repo) (just my opinion)

I already suggest this, but As per @jimmylai suggestion, I have not implemented this functionality.

acharles7 commented on fixit/cli/add_new_rule.py:
```diff
@@ -0,0 +1,88 @@
+   $ python -m fixit.cli.add_new_rule
+   $ python -m fixit.cli.add_new_rule --path fixit/rules/new_rule.py
+
+"""
```
Done

acharles7 commented:
> > As far as I could estimate from writing 2 rules is that, saving the rule to `new_file.py` by default isn't that useful. Maybe it can take a positional argument named rule name and use it on both places (the filename and the class name, so that people wouldn't need to replace the class name, because they might just remove the `Rule` at all which is non-compliant with the style standards of this repo) (just my opinion)
> 
> I already suggest this, but As per @jimmylai suggestion, I have not implemented this functionality.

Will add `--name abcd` later on for
* create filename, `abcd.py`
* class name, `class AbcdRule`

jimmylai commented:
> > As far as I could estimate from writing 2 rules is that, saving the rule to `new_file.py` by default isn't that useful. Maybe it can take a positional argument named rule name and use it on both places (the filename and the class name, so that people wouldn't need to replace the class name, because they might just remove the `Rule` at all which is non-compliant with the style standards of this repo) (just my opinion)
> 
> I already suggest this, but As per @jimmylai suggestion, I have not implemented this functionality.

The class name is in camel case and the file name is in underscore-separated-case. They doesn't necessary to be the same name. So provide another parameter like `--rule` or `--name` for class name makes more sense to me.

isidentical commented:
> The class name is in camel case and the file name is in underscore-separated-case. They doesn't necessary to be the same name. So provide another parameter like --rule or --name for class name makes more sense to me.

Yeah, makes sense (the thing that was in my mind was .split('_') + .title() + .join() + 'Rule')

jimmylai commented on docs/source/build_a_lint_rule.ipynb:
```diff
@@ -121,14 +121,86 @@
   },
+  {
+   "cell_type": "raw",
+   "metadata": {},
```
You need to follow the steps in https://nbsphinx.readthedocs.io/en/0.7.1/raw-cells.html to specify the format as `rst`.
(which generates `"raw_mimetype": "text/restructuredtext"` metadata)
Otherwise the output is in wrong format.
![image](https://user-images.githubusercontent.com/3840867/92533233-debe3180-f1e6-11ea-80a7-e0c407bd96ee.png)
 
You can run `tox -e docs` locally to generate html files and verify the result in a browser.

acharles7 replied:
thank you, looking into it

acharles7 replied:
Format specified

jimmylai commented on docs/source/build_a_lint_rule.ipynb:
```diff
@@ -121,14 +121,86 @@
+    }
+   }
+  },
+  {
+   "cell_type": "raw",
+   "source": [
+    "Use above structure to create a new rule. \n",
+    "\n",
+    "The Declarative Matcher API\n",
+    "===========================\n",
+    "\n",
+    "Once we have a ``ClassDef`` node, we need to see if it contains a base class named ``object``.\n",
+    "We could implement by inspecting attributes of the node using equality and isinstance."
+   ],
+   "metadata": {
+    "collapsed": false
+   }
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# check if any of the base classes of this class def is \"object\"\n",
+    "def visit_ClassDef(self, node: cst.ClassDef):\n",
+    "    has_object_base = any(\n",
+    "        isinstance(arg.value, cst.Name) and arg.value.value == \"object\"\n",
+    "        for arg in node.bases\n",
+    "    )"
+   ]
+  },
```
Those two cells are repeated.

acharles7 replied:
Removed.

acharles7 commented:
> > The class name is in camel case and the file name is in underscore-separated-case. They doesn't necessary to be the same name. So provide another parameter like --rule or --name for class name makes more sense to me.
> 
> Yeah, makes sense (the thing that was in my mind was .split('_') + .title() + .join() + 'Rule')

How about this?
```python
def snake_to_camelcase(name):
  """Convert snake-case string to camel-case string."""
  return "".join(n.capitalize() for n in name.split("_"))
```

jimmylai commented:
> > > The class name is in camel case and the file name is in underscore-separated-case. They doesn't necessary to be the same name. So provide another parameter like --rule or --name for class name makes more sense to me.
> > 
> > 
> > Yeah, makes sense (the thing that was in my mind was .split('_') + .title() + .join() + 'Rule')
> 
> How about this?
> 
> ```python
> def snake_to_camelcase(name):
>   """Convert snake-case string to camel-case string."""
>   return "".join(n.capitalize() for n in name.split("_"))
> ```

Yeah, that works. Then we should explain it clearly in the documentation.

acharles7 commented:
@jimmylai Should I start working on `--name` or Is it for the future?

jimmylai commented on docs/source/build_a_lint_rule.ipynb:
```diff
@@ -211,7 +183,7 @@
   {
    "cell_type": "code",
-   "execution_count": null,
+   "execution_count": 5,
```
`execution_count` is assigned at run time, the value should be `null`.

jimmylai replied:
When any of the cell is already executed, NBSphinx won't run the notebook in CI.
![image](https://user-images.githubusercontent.com/3840867/92536059-91918e00-f1ed-11ea-8089-13e7ed35bb66.png)

acharles7 replied:
Okay, will make sure, NBSphinx new to me, that's why. Thank you for helping me with this.

acharles7 replied:
done

jimmylai commented on docs/source/build_a_lint_rule.ipynb:
```diff
@@ -130,6 +130,50 @@
+    "\n",
+    "This will generate a model rule file used to create new rule."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "pycharm": {
+     "name": "#%%\n"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "import libcst as cst\n",
+    "import libcst.matchers as m\n",
+    "\n",
+    "from fixit import CstLintRule, InvalidTestCase as Invalid, ValidTestCase as Valid\n",
+    "\n",
+    "class Rule(CstLintRule):\n",
+    "    \"\"\"\n",
+    "    docstring or new_rule description\n",
+    "    \"\"\"\n",
+    "\n",
+    "    MESSAGE = \"Enter rule description message\"\n",
+    "\n",
+    "    VALID = [Valid(\"'example'\")]     # Valid examples\n",
+    "\n",
+    "    INVALID = [Invalid(\"'example'\")] # Invalid examples\n",
+    "    ..."
+   ]
+  },
+  {
+   "cell_type": "raw",
+   "metadata": {
+    "raw_mimetype": "text/restructuredtext"
+   },
+   "source": [
+    "Use above structure to create a new rule. \n",
+    "\n",
```
This looks good now.
Sorry for not thinking of it earlier. Do you mind make one more change?
The best way to demonstrate the command is to actually run it in the tutorial instead of hardcoded the result. So the output will be automatically generated and updated whenever the source code changes.

Can you update this doc to actually run the command to add a new rule?
You can run the command and just cat its content by.
```
! python -m fixit.cli.add_new_rule --path fixit/rules/my_rule.py
! cat fixit/rules/my_rule.py
```
https://github.com/Instagram/Fixit/blob/master/docs/source/getting_started.ipynb
https://fixit.readthedocs.io/en/latest/getting_started.html

acharles7 replied:
No worries, I'll add it.

jimmylai commented:
> @jimmylai Should I start working on `--name` or Is it for the future?

We can leave the `name` parameter change as the next PR and keep this PR simpler.

jimmylai commented on docs/source/build_a_lint_rule.ipynb:
```diff
@@ -130,6 +130,50 @@
+    }
+   },
+   "outputs": [],
+   "source": [
+    "import libcst as cst\n",
+    "import libcst.matchers as m\n",
+    "\n",
+    "from fixit import CstLintRule, InvalidTestCase as Invalid, ValidTestCase as Valid\n",
+    "\n",
+    "class Rule(CstLintRule):\n",
+    "    \"\"\"\n",
+    "    docstring or new_rule description\n",
+    "    \"\"\"\n",
+    "\n",
+    "    MESSAGE = \"Enter rule description message\"\n",
+    "\n",
+    "    VALID = [Valid(\"'example'\")]     # Valid examples\n",
+    "\n",
+    "    INVALID = [Invalid(\"'example'\")] # Invalid examples\n",
+    "    ..."
+   ]
+  },
```
The changes you added to Getting Started should just replace this part. So whenever lint rule template changes, we don't need to manually update this example.
Getting Started is not need to be updated.

Before commit change, please click on "Restart & Clear Output". So all output are not stored.
The CI job will run through each cell and generate the result on the fly.
![image](https://user-images.githubusercontent.com/3840867/92777925-878a9f00-f355-11ea-94c3-f7c86529eceb.png)

acharles7 replied:
Got it. I thought, we have to run those cells in order to show output.

jimmylai commented on fixit/cli/add_new_rule.py:
```diff
@@ -0,0 +1,95 @@
+    """Check for valid path, if yes, return `Path` else raise `Error` """
+    filepath = Path(path)
+    if filepath.exists():
+        raise FileNotFoundError(f"{filepath} already exists")
```
FileExistsError is better aligned with the error message.
```suggestion
        raise FileExistsError(f"{filepath} already exists")
```

jimmylai commented:
It's getting closer.
The added commands doesn't work and show this error.
![image](https://user-images.githubusercontent.com/3840867/92803508-57022f80-f36c-11ea-9e9a-e8b0432751cc.png)

It's because the working directory is `docs/source`, rather than the root of project, so it cannot find `fixit/rules`.

Let's change the working directory with a hidden cell.

![image](https://user-images.githubusercontent.com/3840867/92803743-8fa20900-f36c-11ea-9660-ecc413074f93.png)

To hide the `os.chdir` cell, you can click View to enable the Edit Metadata button and then add "nbsphinx": "hidden" to the metadata to hide it. So it won't be included in the generated html.
![image](https://user-images.githubusercontent.com/3840867/92803913-bc562080-f36c-11ea-8ac4-3a574d66a155.png)

Previously, the `add_rule` command is in the same cell as the `cat` which their output will print together.
Let's also separate them as two different cell to be clear.
Lastly, the documentation you added is better for "build_a_lint_rule". Add it to "getting_started" looks redundant. So let's not to add it to "getting_started".

jimmylai commented:
@acharles7 it looks really awesome now!
Thanks for your contribution and sorry for so many back and forth.

![image](https://user-images.githubusercontent.com/3840867/92814596-4061d580-f378-11ea-9a34-5c9d2be477d8.png)

acharles7 commented:
@jimmylai Thank you for your quick reviews. 
Should I start working on `--name`?

jimmylai commented:
> @jimmylai Thank you for your quick reviews.
> Should I start working on `--name`?

Sure, please go ahead.
